src/client/dfuse/pil4dfs/dfs_dcache.c

```diff
@@ -0,0 +1,663 @@
+{
+	d_hash_rec_addref(&dcache->dd_dir_hash, &rec->dr_entry);
+}
+
+static void
+drec_decref_act(dfs_dcache_t *dcache, dcache_rec_t *rec)
+{
+	d_hash_rec_decref(&dcache->dd_dir_hash, &rec->dr_entry);
+}
+
+static void
+drec_del_at_act(dfs_dcache_t *dcache, dcache_rec_t *rec)
+{
+	d_hash_rec_decref(&dcache->dd_dir_hash, &rec->dr_entry);
+	if (!atomic_flag_test_and_set(&rec->dr_deleted))
+		d_hash_rec_delete_at(&dcache->dd_dir_hash, &rec->dr_entry);
+}
+
+static int
+drec_del_act(dfs_dcache_t *dcache, char *path, dcache_rec_t *parent)
+{
+	const size_t  key_prefix_len = DCACHE_KEY_PREF_SIZE - 1;
+	size_t        path_len;
+	char         *bname = NULL;
+	size_t        bname_len;
+	char         *key = NULL;
+	size_t        key_len;
+	d_list_t     *rlink;
+	dcache_rec_t *rec;
+	int           rc;
+
+	D_ASSERT(path != NULL && path[0] == '/' && path[1] != '\0');
+	D_ASSERT(parent != NULL);
+
+	path_len = strnlen(path, PATH_MAX);
+	D_ASSERT(path_len < PATH_MAX);
+	bname = path + path_len - 1;
+	while (bname > path && *bname != '/')
+		--bname;
+	++bname;
+	bname_len = (path_len + path) - bname;
+
+	key_len = key_prefix_len + bname_len;
+	D_ALLOC(key, key_len + 1);
+	if (key == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+	memcpy(key, parent->dr_key_child_prefix, key_prefix_len);
+	memcpy(key + key_prefix_len, bname, bname_len);
+	key[key_len] = '\0';
+
+	rlink = d_hash_rec_find(&dcache->dd_dir_hash, key, key_len);
+	if (rlink == NULL)
+		D_GOTO(out, rc = -DER_NONEXIST);
+
+	rec = dlist2drec(rlink);
+	drec_del_at(dcache, rec);
+	rc = -DER_SUCCESS;
+
+out:
+	D_FREE(key);
+	return rc;
+}
+
+static int
+dcache_destroy_dact(dfs_dcache_t *dcache);
+static int
+dcache_find_insert_dact(dfs_dcache_t *dcache, char *path, size_t path_len, dcache_rec_t **rec);
+static void
+drec_del_at_dact(dfs_dcache_t *dcache, dcache_rec_t *rec);
+
+static int
+dcache_create_dact(dfs_t *dfs, dfs_dcache_t **dcache)
+{
+	dfs_dcache_t *dcache_tmp;
+	int           rc;
+
+	D_ALLOC_PTR(dcache_tmp);
+	if (dcache_tmp == NULL)
+		D_GOTO(error, rc = -DER_NOMEM);
+
+	dcache_tmp->destroy_fn     = dcache_destroy_dact;
+	dcache_tmp->find_insert_fn = dcache_find_insert_dact;
+	dcache_tmp->drec_incref_fn = NULL;
+	dcache_tmp->drec_decref_fn = drec_del_at_dact;
+	dcache_tmp->drec_del_at_fn = drec_del_at_dact;
+	dcache_tmp->drec_del_fn    = NULL;
+
+	dcache_tmp->dd_dfs = dfs;
+
+	*dcache = dcache_tmp;
+	D_GOTO(out, rc = -DER_SUCCESS);
+
+error:
+	D_FREE(dcache_tmp);
+out:
+	return rc;
+}
+
+static int
+dcache_destroy_dact(dfs_dcache_t *dcache)
+{
+	D_FREE(dcache);
+
+	return -DER_SUCCESS;
+}
+
+static int
+dcache_find_insert_dact(dfs_dcache_t *dcache, char *path, size_t path_len, dcache_rec_t **rec)
+{
+	dcache_rec_t *rec_tmp = NULL;
+	dfs_obj_t    *obj     = NULL;
+	mode_t        mode;
+	int           rc;
+
+	(void)path_len; /* unused */
+
+	D_ALLOC_PTR(rec_tmp);
+	if (rec_tmp == NULL)
+		D_GOTO(error, rc = -DER_NOMEM);
+
+	rc = dfs_lookup(dcache->dd_dfs, path, O_RDWR, &obj, &mode, NULL);
```

Comment:
I assume this is the non-caching case here and it's delegating the path walk to dfs?  This is OK for now but in future then we'll want to do this in pil4dfs in future.

Reply:
https://github.com/daos-stack/daos/blob/master/docs/user/filesystem.md#unified-namespace-uns

dfs supports one container but all our other middleware layered on top if it will detect UNS links on directories and open a new container as necessary.  To do that the path walk needs to happen at the higher level, check the xattr on loopup and open new pools/containers as required.  Out of scope for here but this will likely be the next development work on pil4dfs.

Reply:
OK thanks for the explanation.

Reply:
I will add a comment about this in the follow-up PR

Reply:
Fixed in the GC follow-up PR with commit 23ec8eac9a17a2ddc7872bfcbcdaee1df043ac3e